### PR TITLE
[BugFix] Fix LSTMModule padding bug

### LINKED ISSUE #3711: [BUG] LSTMModule Bug
## Describe the bug

In LSTMModule, during the forward pass, the hidden states can be incorrectly removed. We believe this is because `inv_pad_sequence` expects it to be stored at the end of the trajectory step and removes the following steps accordingly, while in reality, the hidden state is stored at the max number of steps of the batch. This means trajectories with fewer steps than the max will see their hidden state incorrectly be zero. 

## To Reproduce

Steps to reproduce the behavior.

Please try to provide a minimal example to reproduce the bug. Error messages and stack traces are also helpful.

Please use the markdown code blocks for both code and stack traces.

```python
import torch
from tensordict import TensorDict
from torchrl.modules import LSTMModule
from torchrl.modules.tensordict_module.rnn import set_recurrent_mode

B = 1  # Batch Size (1 Env)
T = 5  # Time Steps
F = 2  # Observation Features
H = 4  # Hidden Size

is_init1 = torch.zeros(B, T, 1, dtype=torch.bool)
is_init1[0, 3] = True 

obs1 = torch.ones(B, T, F)
obs1[0, 3:] = 2.0

data = TensorDict({
    "obs": obs1,
    "is_init": is_init1
}, [B, T])

lstm = LSTMModule(
    input_size=F, 
    hidden_size=H, 
    in_key="obs", 
    out_keys=["action_features", ("next", "recurrent_state_h"), ("next", "recurrent_state_c")]
)

with set_recurrent_mode(True), torch.no_grad():
    data_recurrent = lstm(data)
print(data_recurrent["next", "recurrent_state_h"])
```

## Expected behavior

I'd expect `data_recurrent["next", "recurrent_state_h"]` to be populated at the steps of the end of each trajectory, but instead we see zeros at the end of trajectory 2.

## System info

Describe the characteristic of your environment:
 * Describe how the library was installed (pip, source, ...)
 * Python version
 * Versions of any other relevant libraries

```python
import torchrl, numpy, sys
print(torchrl.__version__, numpy.__version__, sys.version, sys.platform)
```

## Reason and Possible fixes

We can contribute a PR to fix this, but it's because the hidden states are stored at the max length trajectory location, and then when we remove the padding, we remove the end of the padding for shorter trajectories, removing the stored hidden states. 

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

### PR DESCRIPTION
**Pull request by lin-erica**
## Description

Fixes [bug](https://github.com/pytorch/rl/issues/3711). 

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by vmoens** [comment]
Thanks for the PR!
I gave a stab at this based on your solution with Claude but the code it came up with looks pretty suboptimal (throughput-wise) to me. These nested loops freak me out a little.
I'll refine that tomorrow!
